### DISCUSSION TITLE
How to start the Workflow GLSP Example in VSCode ?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#519] [Q&A] [answered]
I try to understand how VSCode integration works. I have checked out the glsp-server code and started the server. I was already able to run the glsp-client in my web browser.

Next I tried the glsp-vscode-integration.

- I build the *glsp-vscode-integration* project with `yarn install`
- I started my glsp sever `$ java -jar org.eclipse.glsp.example.workflow-0.10.0-SNAPSHOT-glsp.jar org.eclipse.glsp.example.workflow.launch.ExampleServerLauncher`
- I started Visual Studio Code
- I opened my glsp-vscocde-integration repo in VSCode

But what now??

In the README I found the following sentence:

_To test the VSCode extension with an external server (e.g started from your IDE) the launch configuration "Workflow GLSP Example Extension (External GLSP Server)" can be used._

What is meant by "Workflow GLSP Example Extension (External GLSP Server)"? What is this? And how do I start this extension or where can I find this?

Sorry for this stupid question but I am totally lost

Thanks for any help

**Comment by planger** [comment]
> _To test the VSCode extension with an external server (e.g started from your IDE) the launch configuration "Workflow GLSP Example Extension (External GLSP Server)" can be used._
>
> What is meant by "Workflow GLSP Example Extension (External GLSP Server)"? What is this? And how do I start this extension or where can I find this?

This refers to a launch configuration of the VSCode folder:
![image](https://user-images.githubusercontent.com/588090/152489729-e08e5465-62ae-40ed-85d0-704eda81e86d.png)

They are defined in `.vscode/launch.json` of the repository. If you open the repository in VSCode, they will show up in the _Run and Debug_ view as shown above.

**Comment by rsoika** [comment]
Ah! Yes this works. Very cool!
I'm really impressed with this concept

**Comment by rsoika** [comment]
Just one more additional question I did not understand so far.
To support a custom modeller tool for VSCode, Theia and Web, is it necessary to implement the visual implementations of the modelling elements for each platform separately? Or is this be handled by the GLSP framework?
Or asked the other way: do I need to implement my modelling tools 3 times if I plan to support all three platforms?

**Comment by planger** [comment, marked as answer]
A GLSP diagram editor implementation typically consists of three parts:

1. Client based on [glsp-client](https://github.com/eclipse-glsp/glsp-client), e.g. [workflow-glsp](https://github.com/eclipse-glsp/glsp-client/tree/master/examples/workflow-glsp).
2. Server based on [glsp-server (java)](https://github.com/eclipse-glsp/glsp-server) or [glsp-server-node (nodejs)](https://github.com/eclipse-glsp/glsp-server-node), e.g. [org.eclipse.glsp.example.workflow](https://github.com/eclipse-glsp/glsp-server/tree/master/examples/org.eclipse.glsp.example.workflow)
3. Platform integration: e.g. [standalone](https://github.com/eclipse-glsp/glsp-client/tree/master/examples/workflow-standalone), [Eclipse Theia](https://github.com/eclipse-glsp/glsp-theia-integration/tree/master/examples/workflow-theia), [VSCode](https://github.com/eclipse-glsp/glsp-vscode-integration/tree/master/example/workflow), [Eclipse RCP](https://github.com/eclipse-glsp/glsp-eclipse-integration)

If you now want to support multiple platforms, you only need to provide multiple implementations of part 3 above.
All other parts are the same for every platform. The platform integration takes care of things like the editor registration for a specific file type, platform-specific styling and potentially theming support (VSCode and Theia), native menus, integration with the platforms _Problems View_, etc. Thus, they only contain the things that are really specific to the platform.

The platform-independent things, such as the diagram logic (server) and the diagram rendering&UI (client) remain the same across all platforms.

As an example, please look at the workflow example linked above. This diagram implementation can be used for all platforms; that is, for the same workflow-glsp & workflow server components there are four platform-specific integration components.

**Reply by rsoika** [reply]
Thanks for clarification. But still I am not sure if I understand the project structure. 
When you say I need for example for the server component:

[glsp-server (java)](https://github.com/eclipse-glsp/glsp-server)  e.g. [org.eclipse.glsp.example.workflow](https://github.com/eclipse-glsp/glsp-server/tree/master/examples/org.eclipse.glsp.example.workflow)

but the source code of these two servers are part of the same code repository it is a little bit confusing.

I would expect that the [glsp-server (java)](https://github.com/eclipse-glsp/glsp-server) a kind of library which I can use as a dependency for my own [org.eclipse.glsp.example.workflow](https://github.com/eclipse-glsp/glsp-server/tree/master/examples/org.eclipse.glsp.example.workflow). Is this assumption correct?

Or do you mean, that I need to copy also all the plugin code from the glsp-server into my own new project? I can't believe this. 

Why do you not separate the workflow example into its own git repo, so that developers can make a fork of it?
I fear that this is for some reason not possible ....??

Sorry for that kind of stupid questions but it is hard to get started with some kind of silly hello-world example.

**Reply by planger** [reply]
Sure, the glsp-server is a library that can be consumed via [m2 or p2](https://github.com/eclipse-glsp/glsp-server#maven-repositories-). The example just happens to be in the same repository.

There is a minimal example that may help to better investigate the typical project structure: https://github.com/eclipse-glsp/glsp-examples/tree/master/minimal
However it doesn't use EMF on the backend to really keep it minimal. We are working on having a minimal scenario for each technology combination (VSCode, Theia, EMF or no EMF, with model server backend or not).
